api/src/auth/utils/generate-callback-url.ts

```diff
@@ -0,0 +1,46 @@
+import { useEnv } from '@directus/env';
+import { toArray } from '@directus/utils';
+import { Url } from '../../utils/url.js';
+
+/**
+ * Find matching origin from request
+ * @param requestOrigin - The origin of the request
+ * @param allowedOrigins - The allowed origins from AUTH_ALLOWED_ORIGINS
+ * @returns The matching origin or null
+ */
+function findMatchingOrigin(requestOrigin: string, allowedOrigins: string[]): string | null {
+	for (const allowedOrigin of allowedOrigins) {
+		if (!URL.canParse(allowedOrigin)) continue;
+
+		const { protocol, host } = new URL(allowedOrigin);
+		const allowedOriginBase = `${protocol}//${host}`;
+
+		if (requestOrigin === allowedOriginBase) {
```

Comment:
This would match both `www.example.com/api` and `www.example.com/api2`, returning whichever appears first in the list, even when they are configured as separate allowed origins. This could result in the wrong origin being returned.

Reply:
You are correct in regards to them being the same Directus instance, I misread the first point.

While they are the same instance we should not be allowing/redirecting a different subpath than what you sent. Initiating the login on `www.example.com/company1` should not potentially redirect you to a callback with `www.example.com/company2`. On the same note, if you sent `www.example.com/company3` which is NOT in the list it should be defaulting to the public url.